golem-worker-service-base/src/gateway_api_definition/http/http_api_definition.rs

```diff
@@ -50,26 +49,114 @@ pub struct HttpApiDefinition {
 }
 
 impl HttpApiDefinition {
-    pub fn new(
+    pub fn id(&self) -> ApiDefinitionId {
+        self.id.clone()
+    }
+
+    pub fn version(&self) -> ApiVersion {
+        self.version.clone()
+    }
+    pub fn security_schemes(&self) -> Vec<SecuritySchemeReference> {
+        self.routes
+            .iter()
+            .filter_map(|route| {
+                route
+                    .middlewares
+                    .clone()
+                    .and_then(|x| x.get_http_authentication_middleware())
+            })
+            .map(|x| SecuritySchemeReference::from(x.security_scheme))
+            .collect()
+    }
+
+    pub async fn from_http_api_definition_request<Namespace>(
+        namespace: &Namespace,
         request: HttpApiDefinitionRequest,
         created_at: chrono::DateTime<chrono::Utc>,
-    ) -> Self {
-        HttpApiDefinition {
+        security_scheme_service: &Arc<dyn SecuritySchemeService<Namespace> + Send + Sync>,
```

Comment:
Personal opinion, but I don't like passing in services as part of data constructors. At the use-site it will look like the service will become part of the struct, which is not at all what is actually happening.

I think either a top-level function or a method on the SecuritySchemeService would look better here

Reply:
Hmmm. I think it's debatable. A top level function is not easily observable. 
`HtttpApidefinition::from...` is easily seen, and it indicates, to construct it we will need to pass a function that talks about how to get the full scheme given an identifier. I don't see it as a service, but as a higher order function.